### ISSUE TITLE
Destructive crash wiped most of my bookmarks away

### ISSUE THREAD

**Issue by vbertola** [organicmaps/organicmaps#1653]
I was using the app as usual, and then suddenly it crashed, resetting the phone, and when the phone finished booting again and I could reopen the app, all my bookmark sets but four were gone.
Maps.me already had this bug, it happened randomly every now and then, but at least it only deleted the last used bookmark set (it became a zero-sized file) and not almost all of them :-(
Also, Maps.me at least had bookmark backups to restore from...
I don't know if there is any way to help you tracking down this bug - or if there is any chance to get my bookmarks back. I had just updated the app this morning to the new version. My phone is a Huawei P10 running Android/EMUI 9.1.0.290.

**Comment by biodranik** [comment]
Wow, that's a disaster. Did you use an SD card for maps storage? If yes, how old/good is it?  98% of all strange Android crashes happen when maps files on SD card are corrupted because of the faulty SD card.

**Comment by vbertola** [comment]
It could be the SD card, where I store the maps, but if I remember well (at least with Maps.me) bookmarks were stored in the internal memory anyway.
How can I extract the bookmark files in Organic Maps? Are they accessible anywhere?

**Comment by biodranik** [comment]
Yes, if you have root access. Check /data/data/app.organicmaps/files

The app behavior could be undefined if the data on SD card is corrupted. I highly recommend you to replace it with a new one, or at least to stress-test and recheck it and reformat.

**Comment by pastk** [comment]
> I was using the app as usual, and then suddenly it crashed, resetting the phone, and when the phone finished booting again and I could reopen the app, all my bookmark sets but four were gone. Maps.me already had this bug, it happened randomly every now and then, but at least it only deleted the last used bookmark set (it became a zero-sized file) and not almost all of them :-( Also, Maps.me at least had bookmark backups to restore from... I don't know if there is any way to help you tracking down this bug - or if there is any chance to get my bookmarks back. I had just updated the app this morning to the new version. My phone is a Huawei P10 running Android/EMUI 9.1.0.290.

Did it crash after the update to a new version?

**Comment by alcaparra** [comment]
I have also experienced this issue. When my android crashes **while the app is focused**, my bookmarks are wiped clean. Map position and other settings are not affected. I'm trying to reproduce this in a controlled environment to make debugging possible.

**Comment by biodranik** [comment]
Do you store maps on SD?

**Comment by alcaparra** [comment]
I've tried removing the battery while the app is active to simulate a crash but it does not reproduce.

**Comment by biodranik** [comment]
Wow, interesting cases. It would be great if you can reproduce it!

**Comment by biodranik** [comment]
> I switched recently from maps.me to Organic Maps because I lost all my bookmarks in maps.me

> I was using Organic Maps for a few days now and it happened again. I had Organic Maps open and suddenly my phone restarted. When I opened Organic Maps I lost complete folder with all my pins :( I had three folders named Peru, Chile and Bolivia and the one I was using when app crashed is now deleted...

> So I'm sending attached info about my Android phone. Xiaomi Mi 10T Pro 5G
I don't have SD card, only internal phone storage.

> The crash happened shortly after I added new bookmark. I added bookmark and pressed home button and in like 2-3 seconds my phone restarted. I don't know if that could be related but in this folder I had quite many bookmarks, maybe up to one hundred. Is it limited maybe?

> In maps.me it was a little different. At first I noticed that all of the folders was renamed by adding number 111 at the end of the folder name. But everything was working fine so I didn't pay attention to that. I had there maybe 10 to 15 folders. And then happened basically the same as with OM - I was adding bookmark and shortly after that phone restarted but this time in maps.me most of the folders was deleted. I was left only with four folders which I didn't use for some time and all of the newest was deleted. In OM only the recently modified folder was deleted.

**Comment by biodranik** [comment]
On every modification, a previous version of a bookmark file can be saved in a backup location (renamed). When the main file with bookmarks fails to load, OM can automatically load/restore a backed-up version to avoid any similar rare, but painful issues.

CC @kirylkaveryn

**Comment by pastk** [comment]
But bookmarks file is not re-loaded after each write to it, right?

**Comment by biodranik** [comment]
I did not get the question. Bookmark data is loaded into the memory on the startup, and then periodically saved to disk when any change is made.

**Comment by pastk** [comment]
So then its quite likely that the backed up version will be corrupted too.
Its likely that after the first "bad" write and before there is reload (on restart, etc.) there will be more writes that will backup an already corrupted file.

And we can't easily/efficiently check the file after each write. We can do it on reload only.

**Comment by biodranik** [comment]
The chance to corrupt a file during the system-atomic rename operation is very small. The chance of corrupting a file when writing is high.

The current write is done atomically:
1. Write to tmp file
2. std::rename atomically replaces an existing file with a tmp file on iOS and Android.

The corrupted file means that it was corrupted when writing to it. So without an actual check that the written file is invalid, we can not just assume that it will work...

What can be done?
1. Try to load any recently written file (too much overhead for each write).
2. Keep a list of many backups of the same file (OSMand does this).
3. Use some heuristics?
